bcachefs/bcachefs.py

```diff
@@ -131,14 +139,18 @@ def read_file(self, inode: [str, int]) -> memoryview:
         if isinstance(inode, str):
             inode = self.find_dirent(inode).inode
         extents = self._extents_map[inode]
-        file_size = 0
-        for extent in extents:
-            file_size += extent.size
+        file_size = self._inode_map[inode]
+
         _bytes = np.empty(file_size, dtype="<u1")
+
         for extent in extents:
             self._file.seek(extent.offset)
-            self._file.readinto(_bytes[extent.file_offset:
-                                       extent.file_offset+extent.size])
+
+            start = extent.file_offset
+            end = min(extent.file_offset+extent.size, file_size)
+
+            self._file.readinto(_bytes[start:end])
+        
```

Comment:
```suggestion
            self._file.readinto(_bytes[extent.file_offset:
                                       extent.file_offset+extent.size])
```

Sorry I thought that I already commented on this. We can let numpy make the checks to avoid overshooting the indices:

```
>>> import numpy as np
>>> arr = np.arange(10)
>>> arr[5:15]
array([5, 6, 7, 8, 9])
```

Reply:
I dont think we should necessarily rely on numpy for this

Reply:
I'm fairly confident that numpy takes it from python standard convention on lists:

```
>>> arr = list(range(10))
>>> arr[5:15]
[5, 6, 7, 8, 9]
```

In that case, I would think that a lib is in the wrong if it doesn't support this type of checks